servant-server/src/Servant/Server/Internal.hs

```diff
@@ -116,11 +118,10 @@ instance (KnownSymbol capture, FromText a, HasServer sublayout)
      a -> ServerT sublayout m
 
   route Proxy subserver =
-    DynamicRouter $ \ first ->
-      route (Proxy :: Proxy sublayout)
-            (case captured captureProxy first of
-               Nothing  -> return $ failWith NotFound
-               Just v   -> feedTo subserver v)
+    DynamicRouter $ \ first -> case captured captureProxy first of
+       Nothing  -> LeafRouter (\_ r -> r $ Fail err404)
+       Just v   -> route (Proxy :: Proxy sublayout) (feedTo subserver v)
+
```

Comment:
Good point. I'll try to figure out how to delay captures until just before the method check.